source/opt/scalar_analysis_simplification.cpp

```diff
@@ -0,0 +1,436 @@
+  // For instance:
+  //
+  // induction i = 0, i++
+  // temp = i*10
+  // array[i+temp]
+  //
+  // We can fold the i + temp into a single expression. Rec(0,1) + Rec(0,10) can
+  // become Rec(0,11).
+  SENode* FoldRecurrentExpressions(SENode*);
+};
+
+// From a |multiply| build up the accumulator objects.
+bool SENodeSimplifyImpl::AccumulatorsFromMultiply(SENode* multiply,
+                                                  bool negation) {
+  if (multiply->GetChildren().size() != 2 ||
+      multiply->GetType() != SENode::Multiply)
+    return false;
+
+  SENode* operand_1 = multiply->GetChild(0);
+  SENode* operand_2 = multiply->GetChild(1);
+
+  SENode* value_unknown = nullptr;
+  SENode* constant = nullptr;
+
+  // Work out which operand is the unknown value.
+  if (operand_1->GetType() == SENode::ValueUnknown ||
+      operand_1->GetType() == SENode::RecurrentExpr)
+    value_unknown = operand_1;
+  else if (operand_2->GetType() == SENode::ValueUnknown ||
+           operand_2->GetType() == SENode::RecurrentExpr)
+    value_unknown = operand_2;
+
+  // Work out which operand is the constant coefficient.
+  if (operand_1->GetType() == SENode::Constant)
+    constant = operand_1;
+  else if (operand_2->GetType() == SENode::Constant)
+    constant = operand_2;
+
+  // If the expression is not a variable multiplied by a constant coefficient,
+  // exit out.
+  if (!(value_unknown && constant)) {
+    return false;
+  }
+
+  int64_t sign = negation ? -1 : 1;
+
+  // Add the result of the multiplication to the accumulators.
+  if (accumulators_.find(value_unknown) != accumulators_.end()) {
+    accumulators_[value_unknown] +=
+        constant->AsSEConstantNode()->FoldToSingleValue() * sign;
+  } else {
+    accumulators_[value_unknown] =
+        constant->AsSEConstantNode()->FoldToSingleValue() * sign;
+  }
+
+  return true;
+}
+
+SENode* SENodeSimplifyImpl::Simplify() {
+  // We only handle graphs with an addition, multiplication, or negation, at the
+  // root.
+  if (node_->GetType() != SENode::Add && node_->GetType() != SENode::Multiply &&
+      node_->GetType() != SENode::Negative)
+    return node_;
+
+  SENode* simplified_polynomial = SimplifyPolynomial();
+
+  SERecurrentNode* recurrent_expr = nullptr;
+  node_ = simplified_polynomial;
+
+  // Fold recurrent expressions which are with respect to the same loop into a
+  // single recurrent expression.
+  simplified_polynomial = FoldRecurrentExpressions(simplified_polynomial);
+
+  // Traverse the new DAG to find the recurrent expression. If there is more
+  // than one there is nothing further we can do.
+  for (SENode* child : simplified_polynomial->GetChildren()) {
+    if (child->GetType() == SENode::RecurrentExpr) {
+      recurrent_expr = child->AsSERecurrentNode();
+    }
+  }
+
+  // We need to count the number of unique recurrent expressions in the DAG to
+  // ensure there is only one.
+  std::set<SENode*> recurrent_expressions_in_dag;
```

Comment:
Just use a single SENode*. Set it at the first recurrent expression and if you find a later one you have multiple recursions.